src/man/kiwix-search.1
.TH KIWIX-SEARCH "1" "July 2020" "kiwix-tools" "User Commands"
.SH NAME
kiwix-search \- find articles using a fulltext search pattern
.SH SYNOPSIS
\fBkiwix-search\fR [OPTIONS] ZIM PATTERN\fR
.SH DESCRIPTION
.TP
ZIM
ZIM file to search
.TP
PATTERN
Words or parts of words to search for in the ZIM file
.TP
\fB\-s\fR, \fB\-\-suggestion\fR
Suggest article titles based on the PATTERN instead of a fulltext search
.TP
\fB\-V\fR, \fB\-\-version\fR
print software version
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Give details about the search process


Changelog
kiwix-tools 3.1.2
=================

 * Use new threadsafe API of kiwix-lib to do suggestions search.

kiwix-tools 3.1.1
=================

 * Fix compilation on Windows' CI

kiwix-tools 3.1.0
=================

 * [SERVER] Add option to block external links

kiwix-tools 3.0.3
=================

 * [MANAGER] Fix broken --version argument parsing

kiwix-tools 3.0.2
=================

 * New option --version for all tools
 * Remove benchmark.sh file.
 * [DOCKER] Add ability to download a file at container start.
 * [CI] Move to github actions instead of travis.
 * [SERVER] Trust the given library by default.
 * [SERVER] Add shortcut alias for option `--address` and `--nodatealias`

kiwix-tools 3.0.1
=================

 * Fix --nodatealiases inverted logic regression

kiwix-tools 3.0.0
=================

 * Move kiwix-serve implementation in kiwix-lib.

kiwix-tools 2.1.0
=================

 * Fix few compilation errors.

kiwix-serve
-----------

 * Use new api to filter the library.
 * Mobile friendly top bar.
 * Add notag parameter to be able to exclude tags from the zim search.


kiwix-tools 2.0.0
=================

kiwix-manage
-----------

 * Better usage()
 * Adding multiple files bug fix
 * Remove download command.

kiwix-serve
-----------

 * Better usage()
 * Display properly welcome page on 3 columns
 * New welcome page footer "Powered by Kiwix"

kiwix-tools 1.2.1
=================

kiwix-serve
-----------

 * Always use POLL when avaible.

kiwix-tools 1.2.0
=================

 * Remove rpath for installed binaries.

kiwix-serve
-----------

 * New Dockerfile of kiwix-serve
 * New --nodatealiases option
 * Do not use POLL on windows

kiwix-manage
------------

 * Do not show all books if book ids has been provided.
 * Be able to add several zim files in the same time in a library.

kiwix-tools 1.1.0
=================

kiwix-serve
-----------

 * Fix bug about handling of absolute url in old zim file.
 * All the catalog to be searched by tags.

kiwix-tools 1.0.0
=================

 * [CI] Use the new deps archive xz
 * Move version 1.0.0. There is no need to stay in pre 1.0 version.

kiwix-serve
-----------

 * Correctly implement redirection.
   kiwix-serve now return a 302 http status code instead of resolving the
   redirection internally and return the content.


kiwix-tools 0.9.0
=================

 * Update README
 * Update man pages
 * Remove support of external indexes (manage, search, serve)
 * Update build system as we don't use ctpp2 anymore
 * Update to last kiwix-lib API.

kiwix-manage
------------

 * Update usage.


kiwix-tools 0.8.0
=================

kiwix-manage
------------

 * Be able to remove several books from the library in one command.

kiwix-tools 0.7.0
=================

 * Adapt to kiwix-lib new API

kiwix-serve
-----------

 * Dumps only valid books in the opdsfeed.
 * Allow the opds feed to be filtered by lang and paginated.

kiwix-manage
------------

 * Add a download command to download a remote book locally
 * Book are referenced by bookId not index.
 * No more indexType option as it is always XAPIAN.

kiwix-tools 0.6.1
=================

kiwix-serve
-----------

 * Update README.
 * Fix crash when `--library` flag is provided without value.
 * Correctly handle mimetype of file without extension on 64bits.
 * Minor fixes

kiwix-tools 0.6.0
=================

 * remove kiwix-install tool.

kiwix-serve
-----------

 * Improved taskbar #160
 * Fix global page when using the option `--nosearchbar`
 * Return 404 for missing resources
 * Fix compilation for gcc 4.8.

kiwix-manage
------------

 * Returns proper exit code (not always 0)


kiwix-tools 0.5.0
=================

 * Build kiwix-tools setting the RPATH
 * Compile without warning.


kiwix-serve
------------

 * Serve metadata information using the "/meta" url.
 * Serve an OPDS stream of all zim handled by kiwix-serve
   All informations cannot be infer from the zim file itself,
   you should use a library.xml to provide needed information (url, ...)
 * Update kiwix-serve to use the new API of kiwix-lib

kiwix-tools 0.4.0
=================

 * Use gcc-5 on travis.

kiwix-serve
-----------

 * Accept zim file with `&` in the name
 * Do not cache (on client side) global search (as it can depends on the zim
   files handled)
 * Fix HTTP byte range handling. (#91)
 * Fix opening of relative path (#70)
 * Add a small (and hidden) API to do geo search.
 * Better request parsing. (#91)
 * Better handling of invalid request (#116)
 * Various bug fixes (#146, #150, #153, #165, #168, #165)

kiwix-search
------------

 * Add an option `--suggestion` to do suggestion search with
   kiwix-search.(#132)

kiwix-tools 0.3.0
=================

 * Move to C++11

kiwix-serve
-----------

 * Add a global taskbar in the welcome page to search in all zims (#49)
 * Serve the taskbar as css file instead of including it in the html (#68):
   * Better client caching
   * The html encoding is now in the first 1024 bytes and firefox correctly
     detect the encoding
 * Make kiwix-server multi-threaded (#82)
 * Correctly return 404 instead of crashing when request inexistant skin file
   (#83)
 * Correctly respond to bytes-range requests.(#84)
 * Directly respond to first request for a url instead of refusing the first
   connexion
 * Add support to relative url location. (#86)
 * Remove caching (on client side) for the welcome page. (#86)


kiwix-tools 0.2.0
=================

 * Remove indexer tools

kiwix-serve
-----------

 * Correctly fix the deflate data we send over http. (#15)
 * Update in the taskbar (or topbar):
    * Taskbar is responsive (github.com/kiwix/kiwix/issues/336)
    * Force css rules for the taskbar (and not be impacted by content's css)
 * Add `--nolibrarybutton` to hide the library button from the taskbar.
 * Rewrite of the welcome page.


kiwix-installer
---------------

 * Remove indexing functionnality


src/searcher/kiwix-search.cpp
/*
 * Copyright 2011 Emmanuel Engelhart <kelson@kiwix.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU  General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 */

#include <getopt.h>
#include <kiwix/reader.h>
#include <kiwix/searcher.h>

#include "../version.h"

void usage()
{
  cout << "Usage: kiwix-search [OPTIONS] ZIM PATTERN" << endl << endl
       << "  kiwix-search allows one to find articles based on the a fulltext search pattern." << endl << endl
       << "  ZIM is the full path of the ZIM file." << endl
       << "  PATTERN is/are word(s) - or part of - to search in the ZIM." << endl << endl
       << "  -s, --suggestion\tSuggest article titles based on the few letters of the PATTERN instead of making a fulltext search. Work a bit like a completion solution." << endl
       << "  -v, --verbose\t\tGive details about the search process" << endl
       << "  -V, --version\t\tPrint software version" << endl;
  exit(1);
}

int main(int argc, char** argv)
{
  /* Init the variables */
  // const char *indexPath =
  // "/home/itamar/.www.kiwix.org/kiwix/43k0i1j4.default/6d2e587b-d586-dc6a-dc6a-e4ef035a1495d15c.index";
  // const char *indexPath = "/home/itamar/testindex";
  const char* zimPath = NULL;
  const char* search = NULL;
  bool verboseFlag = false;
  bool suggestionFlag = false;
  int option_index = 0;
  int c = 0;

  kiwix::Searcher* searcher = NULL;
  kiwix::Reader* reader = NULL;

  /* Argument parsing */
  while (42) {
    static struct option long_options[]
        = {{"verbose", no_argument, 0, 'v'},
           {"suggestion", no_argument, 0, 's'},
           {"version", no_argument, 0, 'V'},
           {0, 0, 0, 0}};

    if (c != -1) {
      c = getopt_long(argc, argv, "Vvsb:", long_options, &option_index);

      switch (c) {
        case 'v':
          verboseFlag = true;
          break;
        case 'V':
          version();
          return 0;
        case 's':
          suggestionFlag = true;
          break;
      }
    } else {
      if (optind < argc) {
        if (zimPath == NULL) {
          zimPath = argv[optind++];
        } else if (search == NULL) {
          search = argv[optind++];
        } else {
          cout << zimPath << endl;
          usage();
        }
      } else {
        break;
      }
    }
  }

  /* Check if we have enough arguments */
  if (zimPath == NULL || search == NULL) {
    usage();
  }

  /* Try to prepare the indexing */
  try {
    reader = new kiwix::Reader(zimPath);
  } catch (...) {
    /* Cannot open the zimPath, maybe it is a plain old xapian database
     * directory */
  }

  if (reader) {
    searcher = new kiwix::Searcher();
    searcher->add_reader(reader);
  } else {
    cerr << "Unable to search through zim '" << zimPath << "'." << endl;
    exit(1);
  }

  /* Start the indexing */
  if (searcher != NULL) {
    string searchString(search);
    if (suggestionFlag) {
        searcher->suggestions(searchString, verboseFlag);
    } else  {
        searcher->search(searchString, 0, 10, verboseFlag);
    }
    kiwix::Result* p_result;
    while ((p_result = searcher->getNextResult())) {
      cout << p_result->get_title() << endl;
      delete p_result;
    }

    delete searcher;
    delete reader;

    //      kiwix::XapianSearcher::terminate();
  } else {
    cerr << "Unable instanciate the Kiwix searcher." << endl;
    exit(1);
  }

  exit(0);
}
